ocpp_codec/structure.py

```diff
@@ -32,7 +32,7 @@ class ErrorCode(types.SimpleType):
 @dataclass
 class OCPPMessage:
     """Base class every OCPP message should inherit from."""
-    messageTypeId: MessageType
+    messageTypeId: MessageType = field(init=False)  # Let subclasses define that field
```

Comment:
I wonder if we shouldn't simply remove it altogether from the dataclass fields... Like not defining it in `OCPPMessage` at all, and only define it class-wide on `Call`, `CallResult` and `CallError`, without type annotation.

This way, in the `parse_structure` method, we can simply skip it from `raw_data` (`raw_data = raw_data[1:]`) after having used it to get the dataclass we need through `_MSGTYPEID_TO_DATACLASS`.

What do you think? I don't really like the `__post_init__` methods and using `del`.

Reply:
I chose the `__post_init__` way because of a [warning from python's documentation](https://docs.python.org/3.7/library/dataclasses.html?highlight=__post_init#dataclasses.replace) about `init=False` fields:

> Be forewarned about how init=False fields work during a call to replace(). They are not copied from the source object, but rather are initialized in __post_init__(), if they’re initialized at all. It is expected that init=False fields will be rarely and judiciously used. If they are used, it might be wise to have alternate class constructors, or perhaps a custom replace() (or similarly named) method which handles instance copying.

With a second thought though, one should not alter an instance's `messageTypeId` and calling `replace()` on such altered message will simply use the default `messageTypeId` for the new instance which is what we want anyway.